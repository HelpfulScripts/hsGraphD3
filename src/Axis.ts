/**
 * # Axis class
 * 
 */

/** */
import { axisTop }          from 'd3';
import { axisRight }        from 'd3';
import { log as gLog }      from 'hsutil';   const log = gLog('Axis');
import { GraphComponent }   from './GraphComponent'; 
import { GraphCfg }         from './GraphComponent';
import { ComponentDefaults }from './GraphComponent';
import * as def             from './Defaults';
import { ScaleDefaults }    from './Scale';
import { UnitVp, d3Base }   from './Defaults';


export enum Direction {
    horizontal  = 'hor',
    vertical    = 'ver'
}

export interface AxisDefaults extends ComponentDefaults {
    color:      string;
    line:       def.Line;
    tickWidth:  UnitVp;
    tickMarks:  def.Line;
    tickLabel:  def.TextStyle;
}

export interface AxesDefaults extends ComponentDefaults {
    color:      string;
    hor:        AxisDefaults;
    ver:        AxisDefaults;
} 


export class Axes extends GraphComponent {
    private axes: Axis[] = [];

    constructor(cfg:GraphCfg) {
        super(cfg, cfg.baseSVG.select('.axes'));
        let axis;
        this.axes.push(new Axis(cfg, Direction.horizontal));
        this.axes.push(new Axis(cfg, Direction.vertical));
    }

    public get componentType() { return 'axes'; }

    public createDefaults():AxesDefaults {
        return {
            color:  '#000',
            hor:    this.axes[0].createDefaults(),
            ver:    this.axes[1].createDefaults()
        };
    }
    
    public renderComponent() {
        this.axes.forEach(axis => axis.renderComponent());
    }
}


export class Axis {
    private dir: Direction;
    private cfg: GraphCfg;
    private svg: d3Base;

    constructor(cfg:GraphCfg, dir:Direction) {
        this.cfg = cfg;
        this.svg = cfg.baseSVG.select('.axes').append('g');
        this.dir = dir;
        this.svg.classed(`${dir}Axis`, true);
    }

    public createDefaults() {
        return {
            color:      'currentColor',
            line:       def.defaultLine(2),
            tickWidth:  10,
            tickMarks:  def.defaultLine(2),
            tickLabel:  def.defaultText()
        };
    }
    
    public renderComponent() {
        const horScales = this.cfg.scales.hor;
        const verScales = this.cfg.scales.ver;
        const style = this.cfg.defaults.axes[this.dir];
        let axis;
        const margins = (<ScaleDefaults>this.cfg.defaults.scales).margin;
        this.cfg.baseSVG.select('.axes')
            .attr('color', this.cfg.defaults.axes.color);
        this.svg
            .attr('stroke', style.line.color)
            .attr('stroke-width', style.line.width)
            .attr('stroke-opacity', style.line.opacity);

        if (this.dir===Direction.horizontal) {
            axis = axisTop(horScales);
            const yCrossing = Math.max(margins.left, Math.min(verScales(0), this.cfg.viewPort.height-margins.right));
            this.svg.attr("transform", `translate(0, ${yCrossing})`);
        } else {
            axis = axisRight(verScales);
            const xCrossing = Math.max(margins.top, Math.min(horScales(0), this.cfg.viewPort.width-margins.bottom));
            this.svg.attr("transform", `translate(${xCrossing}, 0)`);
        }
        axis.tickSize(style.tickWidth);
        this.svg
            .attr('color', style.color)
            .style('font-family',  style.tickLabel.font.family)
            .style('font-size',   `${style.tickLabel.font.size}px`)
            .style('font-style',  style.tickLabel.font.style)
            .style('font-weight', style.tickLabel.font.weight);
        this.svg.call(axis);
        this.svg.selectAll('text')
            .transition().duration(1000)
            .attr('stroke', style.tickLabel.color || 'currentColor')
            .attr('stroke-width', 1);
        }
}