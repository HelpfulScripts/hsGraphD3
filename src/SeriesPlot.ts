/**
 */

/**  */
import { Log }                  from 'hsutil'; const log = new Log('SeriesPlot');
import { BaseType }             from 'd3';
import { format as d3format}    from 'd3';
import { d3Base, Index }        from "./Settings";
import { defaultLine }          from "./Settings";
import { defaultMarker }        from "./Settings";
import { Label }                from "./Settings";
import { defaultLabel }         from "./Settings";
import { schemes }              from "./Settings";
import { defaultArea }          from "./Settings";
import { Area }                 from "./Settings";
import { Line }                 from "./Settings";
import { Marker }               from "./Settings";
import { GraphCfg }             from "./GraphComponent";
import { DataSet }              from "./Graph";
import { DataVal }              from "./Graph";
import { AccessFn }             from "./Graph";
import { DataRow }              from "./Graph";
import { GraphDimensions }      from "./Graph";
import { setLabel }             from "./Settings";
import { setArea }              from "./Settings";
import { setStroke }            from "./Settings";
import { setFill }              from "./Settings";
import { SystemType }           from './Scales';

export type d3Selection = d3.Selection<BaseType, unknown, BaseType, unknown>; 

export interface SeriesPlotDefaults {
    dims:       SeriesDimensions;
    area:       Area;
    line:       Line;
    marker:     Marker;
    label:      Label;
    transition: boolean;
}

/**
 * specifies the valid dimensions (e.g. 'x' for the x-axis) for a {@link SeriesPlot.SeriesPlot `SeriesPlot`}.
 * For each dimension, the value can be either 
 * - a string that identifies the column name in the data set to use
 * - or a {@link SeriesPlot.ValueFn `ValueFn`} function that returns the data to use.
 * </ul>
 * Optional auxiliary dimensions for all plots:<ul>
 * - `label`?: optional values for item {@link SeriesPlot.SeriesDimensions.label labels}
 * - `popup`?: optional values to show in mouse-over {@link SeriesPlot.SeriesDimensions.popup popups}.
 * - `color`?: optional values to determine {@link SeriesPlot.SeriesDimensions.color marker colors}
 * - `stacked`?: optional {@link SeriesPlot.SeriesDimensions.stacked stack group}. 
 *    Series with the same group will be stacked on each other
 */
export interface SeriesDimensions { 
    [dim:string]: ValueDef; 
    /** 
     * optional, name of label data column, or a function returning a value.
     * See {@link Settings.Label Settings.Label} for configuring labels.
     */
    label?: ValueDef;  
    /** 
     * optional, name of popup data column, or a function returning a value.
     * The popup is triggered when the mouse enters a marker or label area.
     * If omitted, a default popup will configured to show the marker's 
     * abscissa and ordinate names and values.
     * See {@link Popup Popup} for an example.
     */
    popup?: ValueDef;   
    /** 
     * optional, name of color data column, or a function returning a value.
     * The returned color will be used on the marker of that data point. If 
     * omitted, the `series'` color will be used, as defined in the default settinmgs:
     * `graph.defaults.series.<seriesKey>.marker.fill`. 
     * See {@link SeriesPlot.SeriesPlot.d3MarkerColors marker colors} for more details.
     */
    color?: ValueDef;   
    /** 
     * optional stack group. Series with the same group will be stacked on each other. 
     * `stacked` differes from other dimensions in that the `ValueDef` is always a `string`
     * used as a unique identifier for the series to be stacked (and not as a named index into the data set).
     * For example: `stacked:'myStack'` would stack all series with the ID `myStack`. 
     */
    stacked?:     string;
}

/** 
 * Basic `ValueDef` definition, used in {@link SeriesPlot.SeriesPlot `SeriesPlot`}. 
 * - `string`: the name of column in the data set, e.g. `x:'time'`
 * - `number`: a constant value, e.g. `r: 5`
 * - `ValueFn`: a {@link SeriesPlot.ValueFn `user-defined function`} returning the value.
 * - `string[]`: a list of column names, e.g. `['from', 'to']`
 */
export type ValueDef = string|number|ValueFn|string[];

/** 
 * a user-defined function returning the {@link Graph.DataVal `DataVal`} of a data point.
 * The function will be called at runtime once for each row in the data set.
 * @param rowIndex the index of the row in the {@link Graph.DataSet `DataSet's`} rows array. 
 * @return the value of type `DataVal` to use for the row.
 */
export interface ValueFn { (rowIndex:Index): DataVal; }

/**
 * coverts a `DataVal` to a `string`
 * @param val 
 */
export function text(val:DataVal) {
    return typeof val==='number'? val.toFixed(1) : <string>val;
}


/**
 * Abstract base class for all series plot types. It manages
 * - setting the series' defaults
 * - expanding domains to values of this series
 * - the rendering lifecycle for the series
 * - maintaining a stack mechanism that allows series to be stacked on one another
 * To create a series plot, add the desired plot type to the graph:
 * ```
 * graph.add(<type>, {<dim>:<ValueDef>, ...});
 * ``` 
 * where `<dim>` specifies a {@link SeriesPlot.SeriesDimensions series dimension}
 * and {@link SeriesPlot.ValueDef `ValueDef`} provides the data values for the dimension, 
 * as one of:
 * - the name of a column in the data set to use
 * - a constant number
 * - or a user-defined {@link SeriesPlot.ValueFn function}, returning the data to use. 
 */
export abstract class SeriesPlot { 
    /** 
     * a list of data column names used,
     * reflecting the list of column names provided during construction.
     */
    private _dims: SeriesDimensions;

    /** the base svg element to render the component into */
    protected svg: d3Base;
    
    /** the render tree configuration */
    protected cfg: GraphCfg;

    /** the unique series key assigned during cinstruction, used to index the series settings. */
    protected seriesKey:string;

    /** return the GraphDimension of the independent axis */
    protected abscissa:string;

    /** 
     * determines the required type of graph for this lot:  
     * - 'cartesian': plot on a cartesian x/y coordinate system 
     * - 'polar': plot on a polar r/phi coordinate system
     * - 'none': plot on a non-metric graph such as `sankey`
     */
    protected type: SystemType;


    constructor(cfg:GraphCfg, seriesName:string, dims:SeriesDimensions) {
        this.cfg = cfg; 
        this.seriesKey = seriesName;
        this._dims = dims;
        this.type = 'none';
    }
    
    public get key() { return this.seriesKey; }

    public get graphType() { return this.type; }

    protected get dims(): SeriesDimensions { return this._dims; }

    public abstract get dimensions():GraphDimensions;

    protected get defaults():SeriesPlotDefaults { return this.cfg.graph.defaults.series[this.key]; }

    /** 
     * Set the defaults for the series. Called during `addSeries`.
     * */
    public getDefaults(): SeriesPlotDefaults {
        const def:SeriesPlotDefaults = {
            dims:   this._dims,
            line:   defaultLine(5),
            marker: defaultMarker(),
            area:   defaultArea(),
            label:  defaultLabel(),
            transition: true
        };
        def.line.rendered = false;
        def.marker.rendered = false;
        def.area.rendered = false;
        def.label.rendered = false;
        return def;
    }

    /**
     * Returns an accessor function `(row:DataVal[], rowIndex:Index) => DataVal` to access the numeric value in a data row.
     * The type of `v` determines how to access the value: 
     * - If `v` is a function it will be valuated for the provided `row` and `rowIndex` to return the result.
     * - If `v` is a number it will be returned as constant result.
     * - If `v` is a string and contained in `colNames` it specifies the column to index in `row` 
     * @param v the `ValueDef` specifying the value
     * @param colNames a list of names for the coluymns in the `DataSet`
     * @return an accessor function `(row:DataVal[], rowIndex:Index) => DataVal` 
     * that returns a `DataVal` value. The function
     * receives a `DataRow` and the index of the row in the `DataSet` as a parameter. 
     */
    protected accessor(v:ValueDef, colNames:string[], useStack=true):(row:DataRow, rowIndex:Index) => DataVal {
        switch (typeof(v)) {
            case 'function':return (row, rowIndex) => (<ValueFn>v)(rowIndex);
            case 'number':  return () => <DataVal>v;
            case 'string':
            default:        const c = colNames.indexOf(''+v); 
                            // try row.data[c] (as passed by d3.arc), then row[c]
                            return (row) => row['data']? row['data'][c] : row[c];
        }
    }


    /**
     * applies `stroke` and `fill` colors to `items`.
     * Colors are determined by the `color` tag in the series definition as in
     * ```
     * graph.add('column', {x:'item', y:'Mary', color:<colorValue>}); 
     * ```
     * Various coloring modes are available via the `<colorValue>` tag:
     * - `string`: 
     *     1. addresses a data column containing
     *          - `numbers`: used to index the `default.marker.scheme` table
     *          - `strings`: used as direct colors; must be in the format `#rgb` 
     *     2. if not addressing a data column, the value addresses an entry in an
     *        internal color schemes list. Available entries are 
     *          [cat10](https://github.com/d3/d3-scale-chromatic#schemeCategory10),
     *          [blues](https://github.com/d3/d3-scale-chromatic#schemeBlues),
     *          [greens](https://github.com/d3/d3-scale-chromatic#schemeGreens),
     *          [greys](https://github.com/d3/d3-scale-chromatic#schemeGreys),
     *          [oranges](https://github.com/d3/d3-scale-chromatic#schemeOranges),
     *          [purples](https://github.com/d3/d3-scale-chromatic#schemePurples),
     *          [reds](https://github.com/d3/d3-scale-chromatic#schemeReds)
     * - `number`: used to index the `default.marker.scheme` table
     * - `function (rowIndex:Index)=>string` a function, returning a `#rgb` string for the 
     *    provided row index
     * @param items the items being colored
     * @param numRows the number of data rows
     */
    protected d3MarkerColors(items:d3Base, data:DataSet, defaults:SeriesPlotDefaults) {
        function getColor(d:number[], i:number):string  {
            const scheme = schemes[defaults.marker.scheme];
            if (typeof colors === 'function') {
                const color = colors(i);
                return typeof color === 'number'? scheme((color % 10) / 10) : <string>color;
            } else if (typeof colors === 'string') {
                const col = data.colNames.indexOf(colors);
                if (col >= 0) {
                    return d['data']? scheme(d['data'][col]) : scheme(d[col]);
                } else if (schemes[colors]) { 
                    return schemes[colors](i / data.rows.length);
                } else {
                    return colors;
                }
            } else if (typeof colors === 'number') {
                return scheme((colors % 10) / 10);
            } else {
                return '#f00';
            }
        }
        const colors = this.dims.color;
        if (colors) {
            // items.attr('stroke', getColor)
            //      .attr('fill', getColor);
            items.attr('color', getColor);
        }
    } 

    protected d3FillColors(items:d3Base, data:DataSet, defaults:SeriesPlotDefaults) {
        function getColor(d:number[], i:number):string  {
            const scheme = schemes[defaults.marker.scheme];
            if (typeof colors === 'function') {
                const color = colors(i);
                return typeof color === 'number'? scheme((color % 10) / 10) : <string>color;
            } else if (typeof colors === 'string') {
                const col = data.colNames.indexOf(colors);
                if (col >= 0) {
                    return d['data']? scheme(d['data'][col]) : scheme(d[col]);
                } else if (schemes[colors]) { 
                    return schemes[colors](i / data.rows.length);
                } else {
                    return colors;
                }
            } else if (typeof colors === 'number') {
                return scheme((colors % 10) / 10);
            } else {
                return '#f00';
            }
        }
        const colors = this.dims.color;
        if (colors) {
            // items.attr('stroke', getColor)
            //      .attr('fill', getColor);
            items.attr('color', getColor);
        }
    }   

    /**
     * formats the popup string to display
     * @param colNames 
     */
    protected d3RenderPopup(data:DataSet):AccessFn {
        if (this.dims.popup) {
            const format = d3format('.4r');
            const popupAccess = this.accessor(this.dims.popup, data.colNames, false);
            const abscissa = this.dimensions[this.abscissa][0];
            const absAccess     = this.accessor(abscissa, data.colNames, false);
            // if dims x present, reflect its column nname; else use 'index'
            const absName = typeof abscissa === 'string'? abscissa : this.abscissa;
            return (r:DataVal[], i:number) => {
                let val = popupAccess(r,i);
                if (typeof val === 'number') { val = format(val); }
                return typeof this.dims.popup === 'function' ? popupAccess(r,i) :`
                    ${this.dims.popup} = ${val}<br>
                    ${absName} = ${absAccess(r,i)}
                `;
            };
        }
    }


    //---------- lifecylce methods -------------------

    public initialize(svg:d3Base, color?:string): void {
        this.svg = svg.append('g').classed(this.seriesKey, true);
        if (color) { this.svg.style('color', color); }

        const defaults = this.defaults;
        if (defaults.area.rendered) {
            this.svg.append('g').classed('area', true).append('path');
            const area = this.svg.select('.area');
            setArea(area, defaults.area);
        }
        if (defaults.line.rendered) {
            this.svg.append('g').classed('line', true).append('path');
            const line = this.svg.select('.line');
            setStroke(line, defaults.line);
        }
        if (defaults.marker.rendered) {
            this.svg.append('g').classed('markers', true);
            const markers = this.svg.select('.markers');
            setStroke(markers, defaults.marker.stroke);
            setFill(markers, defaults.marker.fill);
        }
        if (defaults.label.rendered) {
            this.svg.append('g').classed('label', true);
            const label = this.svg.select('.label');
            setLabel(label, defaults.label);
        }
    }

    public preRender(data:DataSet): void {}

    public renderComponent(data:DataSet): void {
        const defaults = this.defaults;
        if (defaults.marker.rendered) { this.svg.call(this.d3RenderMarkers.bind(this), data); }
        if (defaults.line.rendered)   { this.svg.call(this.d3RenderLine.bind(this), data); }
        if (defaults.area.rendered)   { this.svg.call(this.d3RenderFill.bind(this), data); }
        if (defaults.label.rendered)  { this.svg.call(this.d3RenderLabels.bind(this), data); }
    }

    public postRender(data:DataSet): void {}

    protected abstract d3RenderMarkers(plot:d3Base, data:DataSet):void;
    protected abstract d3RenderLine(plot:d3Base, data:DataSet):void;
    protected abstract d3RenderFill(plot:d3Base, data:DataSet):void;
    protected abstract d3RenderLabels(plot:d3Base, data:DataSet):void;
}

