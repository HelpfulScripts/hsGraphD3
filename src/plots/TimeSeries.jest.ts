import * as hsGraphD3   from '../';
import * as d3 from 'd3';
import { DataSet }      from '../Graph';
import { SeriesPlotDefaults } from '../SeriesPlot';
import { AxesDefaults } from '../Axis';

let clientWidth = 300;
const root = window.document.createElement("div");
root.style.width = `${clientWidth}px`;
root.style.height = "300px";

d3.selection.prototype.transition   = function(){ return this; };
d3.selection.prototype.duration     = function(){ return this; };
d3.selection.prototype.ease         = function(){ return this; };

const data:DataSet = {
    colNames: 
        ['xval', 'yval'],
    rows : [
        [-1,      15,     15],
        [0.2,      2,     23],
        [3,       28,     30],
        [7.5,     13,      8]
    ]
};

function createGraph(root:any) {
    const graph = new hsGraphD3.GraphCartesian(root);
    graph.isRendered = () => true;
    (<AxesDefaults>graph.defaults.axes).hor.numTicksMinor = 10;
    (<AxesDefaults>graph.defaults.axes).hor.numTicksMajor = 2;
    (<AxesDefaults>graph.defaults.axes).ver.numTicksMinor = 10;
    (<AxesDefaults>graph.defaults.axes).ver.numTicksMajor = 2;
    return graph;
}


describe('TimeSeries', () => {
    let graph:hsGraphD3.GraphCartesian;
    beforeAll(() => {
        graph = createGraph(root);
        graph.addSeries('timeseries', {x:'xval', y:'yval'});
        (<SeriesPlotDefaults>graph.defaults.series.series0).marker.rendered = true;
        (<SeriesPlotDefaults>graph.defaults.series.series0).area.rendered = true;
        graph.render(data);
    });
    it('should plot timeSeries', () =>
        expect(root).toMatchSnapshot()
    );
    it('should support settings changes', () => {
        graph.defaults.axes.color = '#666';
        expect(graph.defaults.axes.color).toBe('#666');
    });
    it(`should have 'timeseries' registered`, () =>
        expect(graph.seriesTypes).toContain('timeseries')
    );
});
