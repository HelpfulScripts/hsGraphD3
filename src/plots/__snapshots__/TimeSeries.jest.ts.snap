// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeSeries should plot timeSeries 1`] = `
<div
  style="width: 300px; height: 300px;"
>
  <svg
    class="baseSVG"
    height="100%"
    preserveAspectRatio="xMinYMin meet"
    viewBox="0 0 1000 700"
    width="100%"
  >
    <g
      class="canvas"
    >
      <rect
        class="graphArea"
        fill="#fff"
        fill-opacity="1"
        height="700"
        rx="0"
        ry="0"
        stroke="#00c"
        stroke-opacity="1"
        stroke-width="1"
        width="1000"
      />
    </g>
    <g
      class="grids"
    >
      <g
        class="hor-major-Grid"
        fill-opacity="0"
        stroke="#444"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="980"
          y1="493.0769230769231"
          y2="493.0769230769231"
        />
        <line
          x1="20"
          x2="980"
          y1="246.9230769230769"
          y2="246.9230769230769"
        />
      </g>
      <g
        class="hor-minor-Grid"
        fill-opacity="0"
        stroke="#eee"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="980"
          y1="690"
          y2="690"
        />
        <line
          x1="20"
          x2="980"
          y1="640.7692307692307"
          y2="640.7692307692307"
        />
        <line
          x1="20"
          x2="980"
          y1="591.5384615384615"
          y2="591.5384615384615"
        />
        <line
          x1="20"
          x2="980"
          y1="542.3076923076923"
          y2="542.3076923076923"
        />
        <line
          x1="20"
          x2="980"
          y1="493.0769230769231"
          y2="493.0769230769231"
        />
        <line
          x1="20"
          x2="980"
          y1="443.8461538461538"
          y2="443.8461538461538"
        />
        <line
          x1="20"
          x2="980"
          y1="394.6153846153846"
          y2="394.6153846153846"
        />
        <line
          x1="20"
          x2="980"
          y1="345.3846153846154"
          y2="345.3846153846154"
        />
        <line
          x1="20"
          x2="980"
          y1="296.15384615384613"
          y2="296.15384615384613"
        />
        <line
          x1="20"
          x2="980"
          y1="246.9230769230769"
          y2="246.9230769230769"
        />
        <line
          x1="20"
          x2="980"
          y1="197.69230769230768"
          y2="197.69230769230768"
        />
        <line
          x1="20"
          x2="980"
          y1="148.46153846153845"
          y2="148.46153846153845"
        />
        <line
          x1="20"
          x2="980"
          y1="99.23076923076917"
          y2="99.23076923076917"
        />
        <line
          x1="20"
          x2="980"
          y1="50"
          y2="50"
        />
      </g>
      <g
        class="ver-major-Grid"
        fill-opacity="0"
        stroke="#444"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="651.2328767123288"
          x2="651.2328767123288"
          y1="690"
          y2="50"
        />
      </g>
      <g
        class="ver-minor-Grid"
        fill-opacity="0"
        stroke="#eee"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="125.2054794520548"
          x2="125.2054794520548"
          y1="690"
          y2="50"
        />
        <line
          x1="256.71232876712327"
          x2="256.71232876712327"
          y1="690"
          y2="50"
        />
        <line
          x1="388.2191780821918"
          x2="388.2191780821918"
          y1="690"
          y2="50"
        />
        <line
          x1="519.7260273972602"
          x2="519.7260273972602"
          y1="690"
          y2="50"
        />
        <line
          x1="651.2328767123288"
          x2="651.2328767123288"
          y1="690"
          y2="50"
        />
        <line
          x1="782.7397260273972"
          x2="782.7397260273972"
          y1="690"
          y2="50"
        />
        <line
          x1="914.2465753424657"
          x2="914.2465753424657"
          y1="690"
          y2="50"
        />
      </g>
    </g>
    <g
      class="axes"
      color="#000"
    >
      <g
        class="horAxis"
        color="currentColor"
        fill="none"
        fill-opacity="0"
        font-family="sans-serif"
        font-size="10"
        font-style="normal"
        font-weight="100"
        stroke="currentColor"
        stroke-opacity="1"
        stroke-width="1"
        text-anchor="middle"
        transform="translate(0, 680)"
      >
        <path
          class="domain"
          d="M20.5,-5V0.5H980.5V-5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(125.7054794520548,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            1
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(257.21232876712327,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(388.7191780821918,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            3
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(520.2260273972602,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(651.7328767123288,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(783.2397260273972,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(914.7465753424657,0)"
        >
          <line
            stroke="currentColor"
            y2="-5"
          />
          <text
            dy="0em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            y="-8"
          >
            7
          </text>
        </g>
      </g>
      <g
        class="verAxis"
        color="currentColor"
        fill="none"
        fill-opacity="0"
        font-family="sans-serif"
        font-size="10"
        font-style="normal"
        font-weight="100"
        stroke="currentColor"
        stroke-opacity="1"
        stroke-width="1"
        text-anchor="start"
        transform="translate(50, 0)"
      >
        <path
          class="domain"
          d="M5,690.5H0.5V50.5H5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,690.5)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,641.2692307692307)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,592.0384615384615)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,542.8076923076923)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,493.5769230769231)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            10
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,444.3461538461538)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,395.1153846153846)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            14
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,345.8846153846154)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            16
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,296.65384615384613)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            18
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,247.4230769230769)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            20
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,198.19230769230768)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            22
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,148.96153846153845)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            24
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,99.73076923076917)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            26
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,50.5)"
        >
          <line
            stroke="currentColor"
            x2="5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            fill-opacity="0"
            stroke="currentColor"
            x="8"
          >
            28
          </text>
        </g>
      </g>
    </g>
    <g
      class="series"
    />
    <g
      class="series0"
    >
      <g
        class="lines"
        fill-opacity="0"
        stroke="currentColor"
        stroke-opacity="1"
        stroke-width="5"
      >
        <path
          d="M-137.8082191780822,370C-137.8082191780822,370,-67.67123287671234,743.3333333333334,20.000000000000007,690C107.67123287671234,636.6666666666666,228.21917808219177,95.12820512820512,388.2191780821918,50C548.2191780821918,4.871794871794876,980,419.2307692307692,980,419.2307692307692"
          transform="translate(0)"
        />
      </g>
      <g
        class="markers"
        fill="#F00"
        fill-opacity="1"
        stroke="currentColor"
        stroke-opacity="1"
        stroke-width="4"
      >
        <circle
          cx="-137.8082191780822"
          cy="370"
          r="10"
          transform="translate(0)"
        />
        <circle
          cx="20.000000000000007"
          cy="690"
          r="10"
          transform="translate(0)"
        />
        <circle
          cx="388.2191780821918"
          cy="50"
          r="10"
          transform="translate(0)"
        />
        <circle
          cx="980"
          cy="419.2307692307692"
          r="10"
          transform="translate(0)"
        />
      </g>
    </g>
  </svg>
</div>
`;
